### PR TITLE
PXC-3666: Session with wsrep_on=0 blocks TOI transactions (5.7)

### PR DESCRIPTION
**Pull request by kamil-holubicki**
https://jira.percona.com/browse/PXC-3666

Problem:
When session with wsrep_on=0 executes transaction that acquires MDL lock
needed by parallel TOI transaction, TOI transaction is blocked.

Cause:
There are a couple of reasons:
1. When wsrep_on is set to 0 thd->wsrep_query_state never is set back to
QUERY_IDLE because change is protected by wsrep_on being 1 condition.
This causes async rollbacker to not be triggered, as the query state
within THD is seen by BF transaction as QUERY_EXEC.
2. The logic responsible for proper victim transaction interruption is
hidden behind if(WSREP(thd)) conditions, so works only if victim thread
has wsrep_on enabled.

Solution:
Implement the behavior we have in PXC 8.0. This needs unconditional
execution of the code responsible for victim transaction eviction.
Additionally autocommit retry logic has been improved to return proper
error code in case of wsrep disabled sessions.

### CODE REVIEW

File: mysql-test/suite/galera/t/galera_bf_abort_wsrep_on.inc [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+SET DEBUG_SYNC="NOW WAIT_FOR open_tables_after_open_and_process_table.reached";
+ALTER TABLE t1 ADD KEY(k);
+
+--connection node_1
+--error $expected_error
+--reap
+
+# todo: this should not be here
+#--error 0,ER_LOCK_DEADLOCK
+#COMMIT;
```
Comment (dutow) [inline mysql-test/suite/galera/t/galera_bf_abort_wsrep_on.inc]: is this still neccessary?

File: mysql-test/suite/galera/t/galera_bf_abort_wsrep_on.inc [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+--connection node_2
+ALTER TABLE t1 ADD KEY(k);
+
+--connection node_1
+--error $expected_error
+--reap
+
+# todo: this should not be here
+#--error 0,ER_LOCK_DEADLOCK
+#COMMIT;
```
Comment (dutow) [inline mysql-test/suite/galera/t/galera_bf_abort_wsrep_on.inc]: same question?

File: sql/sql_parse.cc [hunk 3]
```diff
@@ -7961,7 +7986,7 @@ static void wsrep_mysql_parse(THD *thd, const char *rawbuf, uint length,
     if (WSREP(thd)) {
       /* wsrep BF abort in query exec phase */
       mysql_mutex_lock(&thd->LOCK_wsrep_thd);
-      if (thd->wsrep_conflict_state == MUST_ABORT) {
+      if (thd->wsrep_conflict_state== MUST_ABORT) {
```
Comment (venkatesh-prasad-v) [inline sql/sql_parse.cc]: Please add space around `==`

File: sql/sql_parse.cc [hunk 4]
```diff
@@ -1964,10 +1985,14 @@ bool dispatch_command(THD *thd, const COM_DATA *com_data,
       parser_state.reset(beginning_of_next_stmt, length);
       /* TODO: set thd->lex->sql_command to SQLCOM_END here */
 #ifdef WITH_WSREP
+    if (WSREP_ON) {
       wsrep_mysql_parse(thd, beginning_of_next_stmt, length,
                         &parser_state, false);
-#else
+    } else {
```
Comment (venkatesh-prasad-v) [inline sql/sql_parse.cc]: Here and everwhere in this patch: please adjust this as per other if-else in 5.7.
```
if (...)
{
  ...
}
else
{
  ...
}
```

File: sql/sql_parse.cc [hunk 5]
```diff
@@ -1008,14 +1008,25 @@ bool do_command(THD *thd)
   if wsrep_query_state = QUERY_EXEC then thd is not added to rollback thread
   instead it is left out for do_command and dispatch_command to do an explicit
   rollback. */
-  if (WSREP(thd))
+  if (WSREP_ON)
   {
     mysql_mutex_lock(&thd->LOCK_wsrep_thd);
     wsrep_thd_set_query_state(thd, QUERY_IDLE);
-    if (thd->wsrep_conflict_state==MUST_ABORT)
+    if (thd->wsrep_conflict_state== MUST_ABORT)
```
Comment (venkatesh-prasad-v) [inline sql/sql_parse.cc]: Please add space aroud `==`

File: sql/sql_parse.cc [hunk 6]
```diff
@@ -1848,17 +1860,20 @@ bool dispatch_command(THD *thd, const COM_DATA *com_data,
...
+      /* We let COM_QUIT and COM_STMT_CLOSE to execute even if wsrep aborted. */
+      if (command != COM_STMT_CLOSE && command != COM_QUIT) {
+        my_message(ER_LOCK_DEADLOCK,
+                    "WSREP detected deadlock/conflict and aborted the"
+                    " transaction. Try restarting the transaction", MYF(0));
+        WSREP_DEBUG("WSREP detected deadlock/conflict for SQL: %s",
+                    WSREP_QUERY(thd));
+        mysql_mutex_unlock(&thd->LOCK_wsrep_thd);
+        thd->killed= THD::NOT_KILLED;
+        thd->wsrep_conflict_state= NO_CONFLICT;
```
Comment (venkatesh-prasad-v) [inline sql/sql_parse.cc]: I think we should set `thd->wsrep_conflict_state` under `LOCK_wsrep_thd`

File: mysql-test/suite/galera/t/galera_bf_abort_wsrep_on_disabled.test [hunk 7]
```diff
@@ -0,0 +1,11 @@
+#
+# Test a transaction within session with wsrep_on=0 being aborted
+# by TOI transaction.
+# For wsrep enabled sessions we expect that transactions interrupted while
```
Comment (venkatesh-prasad-v) [inline mysql-test/suite/galera/t/galera_bf_abort_wsrep_on_disabled.test]: I think you meant
```suggestion
# For wsrep disabled sessions we expect that transactions interrupted while
```

File: mysql-test/suite/galera/t/galera_bf_abort_wsrep_on.inc [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+# common setup
+--connect node_1a, 127.0.0.1, root, , test, $NODE_MYPORT_1
+--connection node_1
+CREATE TABLE t1(id INT PRIMARY KEY AUTO_INCREMENT, k INT);
+INSERT INTO t1(k) VALUES (1),(2),(3),(101),(102),(103);
+--eval SET SESSION wsrep_on = $wsrep_on_value;
+
+#
+# Case 1: Abort by local TOI when query in IDLE state
```
Comment (venkatesh-prasad-v) [inline mysql-test/suite/galera/t/galera_bf_abort_wsrep_on.inc:24]: In this we have coverage for `IDLE` and `EXEC`. Can we also add one mode scenario for `QUERY_COMMITTING` state? I think we can make use of the `wsrep_before_certification` in `sql/wsrep_hton.cc` debug_sync_point.

**Review by venkatesh-prasad-v — CHANGES_REQUESTED** [review]
Hi Kamil, thanks for working on this issue. Patch looks good, but I have a few comments/suggestions. Please see them below.

In the commit message,
> When wsrep_on is set to 0 thd->wsrep_query_state never is set back to
QUERY_IDLE because change is protected by wsrep_on being 1 condition.
This causes async rollbacker to not be triggered, as the query state
withing THD is seen by FB transaction as QUERY_EXEC.

1. I think in the last line you mean BF transaction.
2. s/withing/within/

File: mysql-test/suite/galera/t/galera_bf_abort_wsrep_on.inc (hunk 1 above)
Comment (kamil-holubicki) [reply mysql-test/suite/galera/t/galera_bf_abort_wsrep_on.inc]: :)

File: mysql-test/suite/galera/t/galera_bf_abort_wsrep_on.inc (hunk 2 above)
Comment (kamil-holubicki) [reply mysql-test/suite/galera/t/galera_bf_abort_wsrep_on.inc]: fixed

File: sql/sql_parse.cc (hunk 3 above)
Comment (kamil-holubicki) [reply sql/sql_parse.cc]: fixed

File: sql/sql_parse.cc (hunk 4 above)
Comment (kamil-holubicki) [reply sql/sql_parse.cc]: This file is not consistent in this matter. 
Adjusted places touched by this patch.

File: sql/sql_parse.cc (hunk 5 above)
Comment (kamil-holubicki) [reply sql/sql_parse.cc]: fixed

File: sql/sql_parse.cc (hunk 6 above)
Comment (kamil-holubicki) [reply sql/sql_parse.cc]: fixed

File: mysql-test/suite/galera/t/galera_bf_abort_wsrep_on_disabled.test (hunk 7 above)
Comment (kamil-holubicki) [reply mysql-test/suite/galera/t/galera_bf_abort_wsrep_on_disabled.test]: fixed

File: mysql-test/suite/galera/t/galera_bf_abort_wsrep_on.inc (hunk 8 above)
Comment (kamil-holubicki) [reply mysql-test/suite/galera/t/galera_bf_abort_wsrep_on.inc:24]: When wsrep_on=0, the query is not replicated, so it is not entering QUERY_COMMITTING state. However, your comment gave me the hint for improving MTR test by introducing 2 more cases and handling 1 corner case :)

**Comment by kamil-holubicki** [comment]
@venkatesh-prasad-v 
Commit message: I will fix it during squashing before the final merge.

File: mysql-test/suite/galera/t/galera_bf_abort_wsrep_on.inc [hunk 9]
```diff
@@ -9,9 +9,15 @@
 
 # common setup
 --connect node_1a, 127.0.0.1, root, , test, $NODE_MYPORT_1
+--connect node_1b, 127.0.0.1, root, , test, $NODE_MYPORT_1
 --connection node_1
 CREATE TABLE t1(id INT PRIMARY KEY AUTO_INCREMENT, k INT);
 INSERT INTO t1(k) VALUES (1),(2),(3),(101),(102),(103);
+
+# auxilary table for easy synchronization between nodes
```
Comment (venkatesh-prasad-v) [inline mysql-test/suite/galera/t/galera_bf_abort_wsrep_on.inc]: s/auxilary/auxiliary/

**Review by venkatesh-prasad-v — COMMENTED** [review]
Thank you for addressing all the review comments. The patch looks better now.
There is one minor comment from my side. Please address it and squash all the commits.

**Review by venkatesh-prasad-v — APPROVED** [review]
Great work!
